### DISCUSSION TITLE
Modifying interrupt handler for nRF52840-dk

### DISCUSSION THREAD

**Discussion by thvdveld** [tock/tock#4072] [Q&A] [answered]
I'm conducting an experiment where I measure the duration between an interrupt firing and an LED set high. I have 2 LEDs: the first one I set high at the beginning of an interrupt, the second one I set high in a task scheduled on a button press.

I am not sure where to find the interrupt handler of a button of the nRF52840-DK. In this interrupt handler, I need to write the following to set the LED on:
```rust
unsafe { core::ptr::write_volatile((0x50000000 + 0x508) as *mut u32, 13u32); }
```

I'm unfamiliar with how Tock works and wonder if this is possible?

**Comment by lschuermann** [comment, marked as answer]
Tock processes interrupts through a cascade of handlers.

1. Hardware interrupts are handled by a platform-specific assembly routine (the top-half interrupt handler). For Cortex-M4 (ARM v7m), the following handler is invoked on an interrupt:
   https://github.com/tock/tock/blob/587f93a5c135c4a6d35834bf368576059ad372c6/arch/cortex-v7m/src/lib.rs#L146-L216

   As these handlers can preempt any other code running on the system, we need to be very careful around unsynchronized accesses to shared state. While these handlers can and do run some Rust code, this is kept deliberately small.

2. When an interrupt fires while a process is executing, the kernel preempts the process and switches back to the kernel loop.

3. The kernel loop picks up on pending interrupts and calls their bottom-half handlers:

   https://github.com/tock/tock/blob/587f93a5c135c4a6d35834bf368576059ad372c6/kernel/src/kernel.rs#L362-L383

4. Through a cascade of function calls, these handlers can enqueue tasks to be run in a userspace process.

5. Pending userspace tasks do *not* preempt a process. Instead, the process has to reach a "yield-point" (implemented through a system call). This point indicates to the kernel that it is safe to schedule a task callback ("upcall") in the process. Only then does a process handle a task.

All of the above stages seem reasonable measurement targets, so where to insert the GPIO write depends on what exactly you quantify. For an end-to-end latency measurement, inserting this GPIO write into the interrupt handler assembly seems reasonable to me. You might need to make sure that the GPIO peripheral has its clock and power domains enabled, for this write to not cause a fault (not sure whether that's necessary on the nRFs).

**Reply by ppannuto** [reply]
As a small FYI / helper: we provide a [`debug_gpio!()`](https://github.com/tock/tock/blob/305cf638998d98312a644233ca307dfaa35066d0/kernel/src/debug.rs#L41) macro to make tossing lightweight toggles like this in easier.

**Reply by thvdveld** [reply]
Thank you to both of you, I think this answered my question! I will play with this today and tomorrow.
